core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapSingleColumnRowInApply.java

```diff
@@ -0,0 +1,202 @@
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.Assignments.Assignment;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.InPredicate;
+import io.trino.sql.tree.LongLiteral;
+import io.trino.sql.tree.QuantifiedComparisonExpression;
+import io.trino.sql.tree.SubscriptExpression;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.sql.planner.plan.Patterns.applyNode;
+
+/**
+ * Given x::row(t) and y::row(t), converts assignments of the form
+ *
+ * <p><code>x IN (y...)</code> => <code>x[1] IN (y[1]...)</code>
+ *
+ * <p>and</p>
+ *
+ * <p> <code>x &lt;comparison&gt; &lt;quantifier&gt; (y...)</code></p> => <code>x[1] &lt;comparison&gt; &lt;quantifier&gt; (y[1]...)</code></p>
+ *
+ * <p>In particular, it transforms a plan with the following shape:</p>
+ *
+ * <pre>
+ * - Apply x IN y
+ *   - S [x :: row(T)]
+ *   - Q [y :: row(T)]
+ *</pre>
+ *
+ * into
+ *
+ * <pre>
+ * - Project (to preserve the outputs of Apply)
+ *   - Apply x' IN y'
+ *     - Project [x' :: T]
+ *         x' = x[1]
+ *       - S [x :: row(T)]
+ *     - Project [y' :: T]
+ *         y' = y[1]
+ *       - Q [y :: row(T)]
+ * </pre>
+ */
+public class UnwrapSingleColumnRowInApply
+        implements Rule<ApplyNode>
+{
+    private static final Pattern<ApplyNode> PATTERN = applyNode();
+
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<ApplyNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    public UnwrapSingleColumnRowInApply(TypeAnalyzer typeAnalyzer)
+    {
+        this.typeAnalyzer = typeAnalyzer;
+    }
+
+    @Override
+    public Result apply(ApplyNode node, Captures captures, Context context)
+    {
+        Assignments.Builder inputAssignments = Assignments.builder()
+                .putIdentities(node.getInput().getOutputSymbols());
+
+        Assignments.Builder nestedPlanAssignments = Assignments.builder()
+                .putIdentities(node.getSubquery().getOutputSymbols());
+
+        boolean applied = false;
+        Assignments.Builder applyAssignments = Assignments.builder();
+        for (Map.Entry<Symbol, Expression> assignment : node.getSubqueryAssignments().entrySet()) {
+            Symbol output = assignment.getKey();
+            Expression expression = assignment.getValue();
+
+            Optional<Unwrapping> unwrapped = Optional.empty();
+            if (expression instanceof InPredicate) {
+                InPredicate predicate = (InPredicate) expression;
+
+                unwrapped = unwrapSingleColumnRow(
+                        context,
+                        predicate.getValue(),
+                        predicate.getValueList(),
+                        (value, list) -> new InPredicate(value.toSymbolReference(), list.toSymbolReference()));
+            }
+            else if (expression instanceof QuantifiedComparisonExpression) {
```

Comment:
This would change the comparison behavior in case of Quantified comparison predicates right? i.e.

```
trino> create table hive.u_padesai.check_unwrap(a bigint, e row(f bigint));


-- with the new optimizer:

trino> SELECT a FROM check_unwrap WHERE e > ALL (SELECT cast(row(null) as row(f bigint)));
 a 
---
(0 rows)

-- without:

trino> SELECT a FROM check_unwrap WHERE e > ALL (SELECT cast(row(null) as row(f bigint)));
Query 20210506_000910_00003_m62ng failed: ROW comparison not supported for fields with null elements
io.trino.spi.TrinoException: ROW comparison not supported for fields with null elements
        at io.trino.spi.type.RowType.checkElementNotNull(RowType.java:688)
        at io.trino.spi.type.RowType.chainComparison(RowType.java:680)
        at io.trino.$gen.PageFilter_20210506_000915_26.filter(Unknown Source)
        at io.trino.$gen.PageFilter_20210506_000915_26.filter(Unknown Source)
        at io.trino.operator.project.PageProcessor.createWorkProcessor(PageProcessor.java:121)
        at io.trino.operator.FilterAndProjectOperator.lambda$new$0(FilterAndProjectOperator.java:57)
```

I don't suspect any failures for currently working queries though. How do we reason about the equivalence of comparison operators here between `row(T)` and `T`?

Reply:
I wrote an analysis for this but can't find it now. Let me see if I can re-create it. The conclusion was that the operations were equivalent, though :)

Reply:
This is the gist of the analysis for `x = SOME (...)`. It can probably be formalized better, but you get the idea. The same reasoning can be applied for all other comparison operators and quantifiers (ALL vs SOME):

```
Given
    x :: row(t)
    y :: multiset(row(t))    -- i.e., a table

and

    x' = x[1]
    y' = { e[1] for all e in y }

Are x = some y and x' = some y' equivalent?



1. x = some y is true if and only if there exists some e in y such that x = e is true

    x = e => x[1] = e[1] => x[1] = some e[1] in y => x' = some y' is true

2. x = some y is false if and only if for all e in y, x = e is false

    x = e is false => x[1] = e[1] is false => x[1] = e[1] is false for all e in y => x' some y' is false

3. x = some y is null if and only if
   a. there is no e in y such that x = e is true

      x = e => x[1] = e[1] => x' = e[1] is false for all e[1] in y' => there is no e[1] in y' such that x' = e[1] is true

   b. there exists some e in y such that x = e is null

      x = e is null if and only if either

        i. x is null

            => x[1] = null => x[1] = e[1] is null for some e in y => x' = some y' is null

        ii. e is null

           => e[1] is null => x[1] = e[1] is null for some e in y => x' = some y' is null

        iii. x[1] is null or e[1] is null

           => x[1] = e[1] is null for some e[1] in y' => x' = some y' is null
```

Reply:
Yes, I'm working on a separate change to support IN and quantified comparisons with row types. I'll put it up for review soon.